### ISSUE TITLE
[pulsar] Report job errors in parallel mode

### ISSUE THREAD

**Issue by ktmbiome** [zdk123/SpiecEasi#80]
Hi,
I am getting an "Error in li[[j]][[i]] : subscript out of bounds" error (or similar) when I run `spiec.easi`.

My code is:

```
pargs2 <- list(rep.num=50, seed=10010, ncores=20)
net <- spiec.easi(phy.filt, method="mb", verbose=TRUE, pulsar.params=pargs2, lambda.min.ratio=1e-3, nlambda=30, sel.criterion='stars', pulsar.select=TRUE)
```

I have tried many different approaches to determine which option(s) may be causative to the error that I am getting. I think it may partially have something to do with pulsar. When I remove the pulsar arguments, I get a different error regarding having negative-length vectors.

When I switch the above `sel.criterion` value to "bstars" instead, I get "Error in lb.premerge[[j]][[i]] : subscript out of bounds". I searched for "lb.premerge" and it appears to come from the pulsar package. Do you happen to know what pulsar is expecting that my OTU table doesn't have? Is it likely to have something to do with the negative-length issue from above?

Thanks in advance for any thoughts you have; let me know if any additional information would be helpful.

My `phy.filt` object information is below:
```
phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 890 taxa and 184 samples ]
sample_data() Sample Data:       [ 184 samples by 159 sample variables ]
tax_table()   Taxonomy Table:    [ 890 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 890 tips and 888 internal nodes ]
```

**Comment by ktmbiome** [comment]
I decided to try a few more things and this now appears to have been resolved by choosing higher `nlambda` (30 -> 100) and smaller `lambda.min.ratio` (1e-3 -> 1e-5) values. I'm still unsure why these changes helped... Also, are there ways to make these types of errors more meaningful, since it had to do with my choice of parameters?

**Comment by zdk123** [comment]
Hi @ktmbiome I've seen these kinds of errors before, but it devilish to track down since it seems to be some weird interaction between specific datasets, the numerical optimization of neighborhood selection - which it turns out isn't independent of a specific choice of lambda.

I agree the error messages should be more meaningful, but its not intuitive to me either why changing the bottom of the lambda path should affect the stability of the algorithm.

If you should provide a reproducible example to reproduce the error that could go a long way,
thanks!

**Comment by ktmbiome** [comment]
Hi @zdk123 ,
Unfortunately, I don't think I can actually share my data. If you can think of general statistics that might be helpful, though, I'm sure it would be fine to send those along.

It would be interesting to determine approximately how to tune these lambda parameters, since I basically trying a bunch of things until something accidentally worked. However, I totally understand that the pulsar code is less-than-helpful. I did head out to there and read through it, but can't exactly tell what gets fed into the line that ultimately breaks either...

Thanks!

**Comment by zdk123** [comment]
FWIW, I am also the pulsar developer so I can also fix things upstream there as well ;-)

But actually I think the biggest factor is some numerical instability with the MB method of huge, since I rarely see these issues with other packages/method combinations.

You are correct that errors are not handled correctly during parallelization, since errors that occur for distributed nodes get returned as strings by parallel, rather than stopping the computation outright, so while "li[[j]]" is a string rather than a list, and so subsetting no longer works. This I *should be* catching in pulsar, but currently am not.

**Comment by ktmbiome** [comment]
My apologies -- I didn't realize you were the developer of pulsar!

Since I am continuing to see what the bounds of the different parameters are for my dataset, I will try the "glasso" method to see if the same issue crops up. I remember trying it yesterday, but can't remember the exact outcome. I should probably just set all of this up in a loop and send it to our HPC for a few days. If I get any useful information, I will let you know. I may also try to generate synthetic communities with your function to see if I can choose parameters that break it.

**Comment by nearinj** [comment]
Hi everyone,

I recently ran into this issue on a dataset that I have been working on...


I ran this:

`> se_ASV_net <- spiec.easi(network_input, method='mb', lambda.min.ratio=1e-2,
+                          nlambda=100, pulsar.params=list(rep.num=10000, ncores=60))`

And the output I received was this:

`Applying data transformations...
Selecting model with pulsar using stars...
Error in li[[j]][[i]] : subscript out of bounds
In addition: Warning message:
In mclapply(isamp, estFun, fargs = fargs, mc.cores = ncores) :
  scheduled core 1, 35, 42, 36, 34, 28, 8, 14, 23, 39, 41, 13, 38, 9, 47, 24, 45, 58, 32, 55, 56, 6, 10, 22, 19, 27, 59, 49, 25, 5, 17, 4, 40, 50, 60, 52, 20, 3, 44, 33, 16 encountered error in user code, all values of the job will be affected`

(Note this was after running the job for six hours on 60 cores)

I have attached an RDS file of the matrix "network_input" that I used in the below dropbox link. Hopefully this can help track down whats causing this issue as I would love to use the tool on my dataset.

https://www.dropbox.com/s/euuiuy6oh8okp36/network_input.rds?dl=1

**Comment by zdk123** [comment]
Is there a particular reason why you've set `rep.num`? In the StARS paper, this value is 20 and I've only ever gone up to 100. Keep in mind replicates are used to evaluate a single parameter lambda, so it's not like bootstrapping where you need a ton of subsamples for very precise estimates.

Either way, bumping down rep.num <= 59 (even for testing purposes) means you'll be able to run the job completely in parallel on a 60 core machine.

The error message indicates an error in some of the individual jobs (1, 35, 42, ...) so the 'subscript out of bounds' means the data was never reduced into the list (there's no data there).

This issue was never really solved upstream and so persists. I'll open up an issue on the pulsar side.

**Comment by nearinj** [comment]
Hi,

Thanks for the clarifications! I was going through a tutorial that mentioned you would want to set the rep number higher in practical settings but I wasn’t sure how high to set it. Thanks for that clarification! I’m testing out the glasso method on the same dataset to see if that passes through without errors. Appreciate the quick responses (:

**Comment by nearinj** [comment]
I re-tested the command:

```
> se_ASV_net <- spiec.easi(network_input, method='mb', lambda.min.ratio=1e-2,
+                          nlambda=100, pulsar.params=list(rep.num=50, ncores=60))
Applying data transformations...
Selecting model with pulsar using stars...
Error: 
101 jobs failed with: "Error in huge.mb(x, lambda = lambda, nlambda = nlambda, lambda.min.ratio = lambda.min.ratio,  : 
  negative length vectors are not allowed"

```

It looks like the update did give a more informative error message! Note that I also tested the glasso method and it did pass without errors.

**Comment by zdk123** [comment]
@nearinj I agree that error message isn't all that expressive, but I was able to pin it down to `huge` package. The good news is that I can reproduce the error on purely synthetic data.  I will open an issue on the huge github.

**Comment by nearinj** [comment]
@zdk123  awesome thanks for following up on this! For now I will just go with using the glasso method as it seems to complete without any issues. :+1:

**Comment by zdk123** [comment]
@nearinj can you check that the issue is resolved after upgrading to huge v1.3.3?

**Comment by nearinj** [comment]
Updating to huge v1.3.3 has indeed solved the issue! 

Thanks for looking into this!
